### ISSUE TITLE
Many lib-runtime-events tests fail on armel

### ISSUE THREAD

**Issue by glondu** [ocaml/ocaml#13234]
Many tests of OCaml 5.2.0 fail on Debian sid, armel.

Quoting the [build log](https://buildd.debian.org/status/fetch.php?pkg=ocaml&arch=armel&ver=5.2.0-1%7Eexp1&stamp=1718265066&raw=0):
```
[...]
List of failed tests:
    tests/lib-runtime-events/'test.ml' with default (bytecode) Process 4069629 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml.ml' with default (bytecode) Process 4069634 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_counters.ml' with default (bytecode) Process 4069639 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_exception.ml' with default (bytecode) Process 4069644 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_parallel.ml' with default (bytecode) Process 4069649 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_reentry.ml' with default (bytecode) Process 4069661 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_runparams.ml' with default (bytecode) Process 4069666 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_slot_reuse.ml' with default (bytecode) Process 4069671 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_caml_stubs_gc.ml' with default (bytecode) Process 4070193 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_compact.ml' with default (bytecode) Process 4070198 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_dropped_events.ml' with line 9 (bytecode) Process 4070203 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_env_start.ml' with default (bytecode) Process 4070211 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_fork.ml' with line 6 (bytecode) Process 4070228 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_user_event.ml' with default (bytecode) Process 4070238 got signal 11(Segmentation fault), no core dumped
    tests/lib-runtime-events/'test_user_event_unknown.ml' with line 7 (bytecode) Process 4070251 got signal 11(Segmentation fault), no core dumped

Summary:
  2480 tests passed
   659 tests skipped
    15 tests failed
   602 tests not started (parent test skipped or failed)
     0 unexpected errors
  3756 tests considered
#### Something failed. Exiting with error status.
```

It can be reproduced with qemu-user-static and an armel chroot.

**Comment by ghost** [comment]
I can't reproduce this on armel on a non-Linux platform, so this is likely something specific to Linux (or Debian, or the specific gcc+glibc version you're using).

Is there a place I can download a Debian/armel image suitable for use with qemu in order to investigate further?

**Comment by glondu** [comment]
I don't have a Debian/armel image at hand, but from Debian/amd64 you can do, after having installed pbuilder, qemu-user-static (and maybe other packages):

```
sudo pbuilder create --basetgz /var/cache/pbuilder/sid-armel.tgz --distribution sid --debootstrapopts --variant=minbase --architecture armel
sudo pbuilder login --basetgz /var/cache/pbuilder/sid-armel.tgz
```

**Comment by ghost** [comment]
> I don't have a Debian/armel image at hand, but from Debian/amd64 you can do, after having installed pbuilder, qemu-user-static (and maybe other packages):
> 
> ```
> sudo pbuilder create --basetgz /var/cache/pbuilder/sid-armel.tgz --distribution sid --debootstrapopts --variant=minbase --architecture armel
> sudo pbuilder login --basetgz /var/cache/pbuilder/sid-armel.tgz
> ```

Unfortunately, I've not been able to make this work (so far) from an ubuntu system (using `--mirror http://www-ftp.lip6.fr/pub/linux/distributions/debian/` as extra argument), `pbuild` complains about being unable to make the `/{bin,lib,sbin}` symlinks to `/usr/{bin,lib,sbin}`. Probably some ubuntu-specific change to `pbuilder` getting in the way. I'll try to get access to a debian setup soonish and retry.

**Comment by lthls** [comment]
I can reproduce the problem with the setup above, on trunk.
However `gdb` doesn't work in this setup so I'm a bit stuck.
I did manage to find out that for `testsuite/tests/lib-runtime-events/test.ml` it segfaults somewhere inside `get_event_counts`, but that's about as far as I could get.

**Comment by lthls** [comment]
With some additional printf debugging I managed to pinpoint to this line:
https://github.com/ocaml/ocaml/blob/1d7d91556fd3a59a086bfd73b7a4c75cfb0ccc7f/otherlibs/runtime_events/runtime_events_consumer.c#L395

**Comment by glondu** [comment]
An alignment issue?

**Comment by glondu** [comment]
According to the [NetBSD manpage](https://man.netbsd.org/atomic_load_acquire.9), `atomic_load_acquire` on 64-bit values might not work on 32-bit architectures when `__HAVE_ATOMIC64_LOADSTORE` is not defined, which is the case here.

**Comment by glondu** [comment]
My observations:
- in the given context (line 395 of runtime_events_consumer), the pointer given to `atomic_load_acquire` is 64-bit-aligned (always 0x424da040 in my setup)
- the segfault happens in `atomic_load_acquire`
- the segfault stays if I remove the call to `atomic_load_acquire`, i.e. if I use directly `runtime_events_buffer_header->ring_head` (looking at the assembly, an atomic load is still done)
- if I copy `sizeof(struct runtime_events_buffer_header)` bytes from `runtime_events_buffer_header` to a freshly `malloc`-ed `foo` (always 0x467e50 in my setup), and use `foo->ring_head`, there is no segfault

It looks like atomic loads are impossible in the memory nearby 0x424da040, but possible nearby 0x467e50. Is that even possible?

**Comment by lthls** [comment]
A wild guess: is it possible that 0x424da040 comes from `mmap`ed memory, and the processor can't perform atomic accesses on such memory ?

**Comment by glondu** [comment]
Right! I could come up with the following reduced repro case:
```
$ dd if=/dev/zero of=test.bin bs=1 count=64
$ cat test.c
#include <stdio.h>
#include <stdint.h>
#include <stdatomic.h>
#include <unistd.h>
#include <sys/mman.h>
#include <fcntl.h>

#define atomic_load_acquire(p)                    \
  atomic_load_explicit((p), memory_order_acquire)

int main() {
  int fd = open("test.bin", O_RDONLY, 0);
  uint64_t *a = mmap(NULL, 64, PROT_READ, MAP_SHARED, fd, 0);
  uint64_t b = atomic_load_acquire(a);
  printf("%lld\n", b);
  return 0;
}
$ gcc test.c -latomic
$ ./a.out
Segmentation fault (core dumped)
```

**Comment by gasche** [comment]
Thanks for the detective chase. I think that we have two choices:
1. try to adapt runtime-events to have a 32bit version manipulating 32bit words rather than 64bit words (so: give up on the current same-data-on-all-archs design), to get it working for armel
2. give up on runtime-events on armel, have a configure check that forbids this combination, and correspondingly disable the testsuite

(2) is of course vastly less work and it is not clear that anyone would care to use runtime-events on armel enough to justify (1). This sounds rather clear-cut?

**Comment by glondu** [comment]
FYI, all tests pass after replacing `atomic_load_acquire` by `memcpy` (3 occurrences). But this is probably not what we want...

> give up on runtime-events on armel, have a configure check that forbids this combination, and correspondingly disable the testsuite

This seems reasonable.

**Comment by shindere** [comment]
> 2. give up on runtime-events on armel, have a configure check that
> forbids this combination, and correspondingly disable the testsuite

Am I understanding correctly that runtime-events should then actually be
disabled on any 32bitsplatform, not jsut armel?

**Comment by lthls** [comment]
> Am I understanding correctly that runtime-events should then actually be disabled on any 32bitsplatform, not jsut armel?

It's likely that the current issue is specific to armel, so I think we should keep support for the other 32-bit platforms unless we get further bug reports.

**Comment by gasche** [comment]
I wonder if other parts of the runtime suffer from the same issue. We use mmap liberally in the runtime (it is used for the minor heap, for small-size pools in the major heap, and for the call stacks in some configurations), but our atomic accesses are typically word-sized. The only other joint use of int64_t and atomics I found is in fiber.c, the `static _Atomic int64_t fiber_id = 0;` global declaration (operated with `fetch_add`), I suppose this is safe?

**Comment by glondu** [comment]
Here are build logs for all Debian architectures and ports:

  https://buildd.debian.org/status/package.php?p=ocaml&suite=experimental

The runtime-events tests seem to pass on other 32-bit architectures (armhf, i386, hppa, hurd-i386, powerpc, sh4 and x32).

I suspect ppc64 will pass if I remove it from native architectures.

The remaining unexplained failure is m68k... a segfault in the first call to ocamlrun... but this probably deserves its own bug report...

**Comment by shindere** [comment]
Vincent Laviron (2024/06/18 07:01 -0700):
> It's likely that the current issue is specific to armel, so I think we
> should keep support for the other 32-bit platforms unless we get
> further bug reports.

Okay I can submit a PR for that, likely tomorrow morning.

**Comment by ghost** [comment]
Before one starts disabling things, I would like to get more information.

Recent enough 32-bit arm processors (v6 and v7) are able to perform 64-bit atomic operations using the `ldrexd` and `strexd` instructions. If gcc knows this (from `-march=` or `-mcpu=` options), it will emit these instructions. Otherwise, it will emit calls to `libatomic` functions (in this case, `__atomic_load_8`); that function in turn will either perform the expected `ldrexd` if supported, or it will invoke a Linux kernel helper function exposed to userland, to trap into the kernel and simulate an atomic (from userland point of view) operation. This helper is only available when the kernel is configured with `CONFIG_KUSERS_HELPERS` is enabled. But then if you are running an SMP kernel, which appears to be the case according to the build log, this helper actually performs an `ldrexd` instruction, which could explain the failure.

@glondu, could you run your simple testcase from [this comment](https://github.com/ocaml/ocaml/issues/13234#issuecomment-2176079951) under gdb and report the results of `x/i $pc` and `info reg` when the program crashes?

**Comment by glondu** [comment]
FYI, armel is armv5: https://wiki.debian.org/ArchitectureSpecificsMemo#armel

**Comment by glondu** [comment]
> @glondu, could you run your simple testcase from [this comment](https://github.com/ocaml/ocaml/issues/13234#issuecomment-2176079951) under gdb and report the results of `x/i $pc` and `info reg` when the program crashes?

```
(sid_armel-dchroot)glondu@abel:~/debug0$ gdb ./a.out
GNU gdb (Debian 13.2-1+b2) 13.2
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./a.out...
(gdb) run
Starting program: /home/glondu/debug0/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabi/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0xffff0f7c in ?? ()
(gdb) x/i $pc
=> 0xffff0f7c:                  @ <UNDEFINED> instruction: 0x01a23f96
(gdb) info reg
r0             0x0                 0
r1             0x0                 0
r2             0xb6fd1000          3070038016
r3             0x0                 0
r4             0x0                 0
r5             0x0                 0
r6             0x0                 0
r7             0x0                 0
r8             0x400634            4195892
r9             0xb6ffece8          3070225640
r10            0x0                 0
r11            0xbefff45c          3204445276
r12            0xb6fc7478          3069998200
sp             0xbefff3f8          0xbefff3f8
lr             0xb6fc74a8          -1224969048
pc             0xffff0f7c          0xffff0f7c
cpsr           0x60000010          1610612752
fpscr          0x0                 0
```

**Comment by ghost** [comment]
> FYI, armel is armv5: https://wiki.debian.org/ArchitectureSpecificsMemo#armel

Oh, that might explain why I couldn't reproduce, then...

**Comment by ghost** [comment]
> > @glondu, could you run your simple testcase from [this comment](https://github.com/ocaml/ocaml/issues/13234#issuecomment-2176079951) under gdb and report the results of `x/i $pc` and `info reg` when the program crashes?
> 
> ```
> (sid_armel-dchroot)glondu@abel:~/debug0$ gdb ./a.out
> Program received signal SIGSEGV, Segmentation fault.
> [...]
> 0xffff0f7c in ?? ()

That address is in the kernel helper code for atomic operations (it's `__kuser_cmpxchg64+0x1c`).

> (gdb) x/i $pc
> => 0xffff0f7c:                  @ <UNDEFINED> instruction: 0x01a23f96

That's `strexdeq r3, r6, [r2]`. So the kernel you are running is compiled with `CONFIG_CPU_32v6K` enabled and emits 64-bit exclusive load/stores.

Note that the program stops with `SIGSEGV`, not `SIGILL`, so it's likely that the instruction is supported. That would hint that memory obtained from `mmap` does not have the right coherency attributes indeed.

**Comment by ghost** [comment]
That said, it looks to me that this is a gcc (libatomic) and/or kernel configuration problem and not an OCaml problem we're seeing here.

**Comment by glondu** [comment]
I've pointed the Debian arm porters to this discussion: https://lists.debian.org/debian-arm/2024/06/msg00002.html

**Comment by glondu** [comment]
I am still interested by the PR @shindere proposed to submit, though.

**Comment by glondu** [comment]
> I am still interested by the PR @shindere proposed to submit, though.

I've implemented one myself: https://salsa.debian.org/ocaml-team/ocaml/-/blob/d848f2f6379179e6fa8ebfdc5a6c69474ad44223/debian/patches/0010-Make-runtime_events-optional.patch

Explicitly disabling runtime_events is left as an exercise to the `./configure` caller.

**Comment by glondu** [comment]
I got an interesting explanation with a patch: https://lists.debian.org/debian-arm/2024/06/msg00004.html

It looks like runtime_events could be supported on armel, at the price of having write access to the mmap-ed file. Is that correct / acceptable in this context? Do we want to keep read-only access elsewhere?

**Comment by ghost** [comment]
I stand by my opinion that this is a toolchain problem.

Clang on arm will either emit an `ldrex` + `clrex` sequence if supported by the target processor, otherwise it will emit a call to `__atomic_load_8` in compiler-rt (its equivalent of `libgcc`) to perform the operation in a slower way, and the implementation makes sure not to perform a write when the operation is an atomic load.

For Debian, if there is a clang package available on this platform, then it might be as simple as mark the OCaml package as requiring clang as its compiler, if the package build infrastructure supports this.

**Comment by wookey** [comment]
It's not a compiler problem - this issue is intrinsic to the architecture. Clang is emitting the right code to do atomic reads, or calling the right atomic function but the mmapped memory it's doing them on has to be marked R/W on armel because it's impossible to do a 64bit atomic read on arm v5 or v6 hardware (i.e avoid the risk of tearing) without writing the value back (and armel can run on v5 and v6). It doesn't matter whether it is the kernel, gcc's atomic functions or clang's atomic functions doing the job - they all have to deal with the same hardware constraints.
The same issue arises for 128-bit atomic reads on arm64. You don't see this problem on armhf because that uses the v7 ISA and thus cannot be run on pre-v7 hardware so the memory doesn't need to be marked R/W.

**Comment by ghost** [comment]
Which is why the compiler-rt implementation uses a slow implementation, which performs two regular 32-bit reads, but with other locks which guarantee atomicity against other threads in the same process. Since the targeted architecture is not SMP capable, this is good enough from a "C11 code in userland" point of view.

**Comment by wookey** [comment]
OK. Yes implementing 'some other lock' is a perfectly reasonable approach. Which could be done explicitly by the software for this arch, but it's nice if it's done automatically by the compiler. And if only clang has this functionality (is that really the case?) then just depending on that to build makes sense. Documenting why this was done would be good, because armel will probably be relegated to 'ports' status in debian quite soon (i.e not a supported release architecture), or ocaml could decide to drop it, or gcc could gain the same functionality, in which case the need to depend on clang would evaporate.

**Comment by glondu** [comment]
Which version of clang supports atomics? The most recent available in Debian testing is 17.

`CC=clang-17 ./configure` fails with:
```
checking whether the C compiler supports _Atomic types... no
configure: error: in `/home/glondu/ocaml-5.2.0':
configure: error: C11 atomic support is required, use another C compiler
```

**Comment by ghost** [comment]
C11 atomics are supported since clang 3.x IIRC. Can you share your `config.log`? Or maybe invoking configure with `CC="clang-17 -std=c11"` would help.

**Comment by glondu** [comment]
```
configure:15343: clang-17 -std=c11 -o conftest -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/glondu/ocaml-5.2.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2 -Wl,-z,relro -Wl,-z,now conftest.c -lm   >&5
conftest.c:59:7: warning: large atomic operation may incur significant performance penalty; the access size (8 bytes) exceeds the max lock-free size (0  bytes) [-Watomic-alignment]
   59 |       atomic_store_explicit(&n, 123, memory_order_release);
      |       ^
/usr/lib/llvm-17/lib/clang/17/include/stdatomic.h:138:31: note: expanded from macro 'atomic_store_explicit'
  138 | #define atomic_store_explicit __c11_atomic_store
      |                               ^
conftest.c:60:9: warning: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Watomic-alignment]
   60 |       * atomic_exchange(&p, 0) = 45;
      |         ^
/usr/lib/llvm-17/lib/clang/17/include/stdatomic.h:143:42: note: expanded from macro 'atomic_exchange'
  143 | #define atomic_exchange(object, desired) __c11_atomic_exchange(object, desired, __ATOMIC_SEQ_CST)
      |                                          ^
conftest.c:61:14: warning: large atomic operation may incur significant performance penalty; the access size (8 bytes) exceeds the max lock-free size (0  bytes) [-Watomic-alignment]
   61 |       return atomic_load_explicit(&n, memory_order_acquire);
      |              ^
/usr/lib/llvm-17/lib/clang/17/include/stdatomic.h:141:30: note: expanded from macro 'atomic_load_explicit'
  141 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
3 warnings generated.
/usr/bin/arm-linux-gnueabi-ld: /tmp/conftest-a248a4.o: in function `main':
././conftest.c:59:(.text+0x3c): undefined reference to `__atomic_store_8'
/usr/bin/arm-linux-gnueabi-ld: ././conftest.c:60:(.text+0x4c): undefined reference to `__atomic_exchange_4'
/usr/bin/arm-linux-gnueabi-ld: ././conftest.c:61:(.text+0x60): undefined reference to `__atomic_load_8'
clang-17: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by ghost** [comment]
Bummer. The clang packages on Debian do not build and provide `libcompiler_rt`. The `configure` step could succeed by adding `-latomic` to the value of `CC` (this is done automatically in `configure` when the compiler is `gcc`). But that would not solve the problem, since the implementation of `__atomic_load_8` in `libatomic` will not work on read-only memory.

Until Debian packages `libcompiler_rt` with `clang` (or as an extra package), this platform will not be able to use the runtime-events library.

Sorry for pointing you in a non-working direction.

**Comment by glondu** [comment]
And what about mmaping with RW permissions?

**Comment by gasche** [comment]
I like the extra protection of read-only memory by default, but I think it would be reasonable to have an ifdef that uses a R/W mapping for armel only, with a comment that explains the reason for this restriction.

**Comment by arndb** [comment]
I don't think that a compile-time check is useful here, as the machine that the code is built on for armel is usally not the one it ends up running on later. The check could go wrong either way: on debian build hosts using arm64, it uses the kernel's cmpxchg64 helper from libatomic that needs write access, but it could do a native atomic load that doesn't with changes to libatomic. On 32-bit machine it might detect that it doesn't need it but then later run on one that does.

Hardcoding based on the target architecture being lower than armv6k should be more reliable.

**Comment by xavierleroy** [comment]
I have the impression that this is fixed by #13267.
